### ISSUE TITLE
[autoscaler] Better documentation on the node provider interface 

### ISSUE THREAD

**Issue by DmitriGekhtman** [ray-project/ray#26087]
### Description

The NodeProviderInterface is public, meant for use in constructing custom Ray deployment stacks.
However, the contracts of the interface aren't clearly documented. That should be fixed.

### Link

See e.g.
https://ray-distributed.slack.com/archives/CN2RGCHRR/p1655932089202649

**Comment by TingkaiLiu** [comment]
Hi! I am the one who asked that question on Slack and also willing to contribute to the documentation / developer guide. Just let me know if I can help.

**Comment by DmitriGekhtman** [comment]
I think we can probably start with more info in the interface file itself.

Some pointers and gap I can think of.

- After terminate_node or create_nodes returns, non_terminated_nodes should be able to read the change immediately.
- After set_node_tags return, node_tags should be able to read the change immediately.
- NodeProvider is called from multiple threads so the implementation should be thread-safe.
- The "nodes" should be virtual machines that can be ssh-d into using the output of the internal_ip and external_ip methods.
  * SSH keys have to provided in the provider config somewhere (provider["auth"] maybe?)
  * The actual SSH command that is executed is somewhere in the logic of `autoscaler/_private/command_runner.py`
- The bootstrap_config method can be used for auto-configuration after the config is submitted. It can in principle be made to mutate the cluster config in arbitrary ways.   
- Don't implement `create_node_with_resources`. That's an experimental method for specialized use-cases.
- The type annotation on the methods internal_ip and external_ip are wrong. It says `str` but should be `Optional[str]`. A `None` return means the node does not yet have an ip assigned. NodeUpdater logic will retry until it has an IP assigned.
- The behavior of the classes NodeUpdater and CommandRunner need to be documented in general, since these are implicit parts of the interface.

Anything else you run into can help inform the docs. If you have time to contribute a NodeProvider developer guide as you work out your implementation, that would be great.
If you have any questions, we can use this thread to discuss.

cc @wuisawesome @ericl

**Comment by wuisawesome** [comment]
that seems like a good start to me.

**Comment by TingkaiLiu** [comment]
Another issue related to IP address (internal_ip()/external_ip() function) after I made create_node() unblocking. 

In my platform, a pending node is usually caused by the lack of resources in my cluster. As a result, the pending Ray worker node is not assigned to any of the physical node and doesn't have an IP address---I have to return None for IP address query. However, the IP address information is not fully updated after the node is up and running.  In the autoscaler monitor the IP of those nodes are still shown as "None", although those node can successfully scale down after max idle time. 

So the question would be: what is the expected behavior of IP query in node_provider for the nodes that don't have IP yet?

**Comment by DmitriGekhtman** [comment]
> The behavior I observed when the create_node() function is blocking:
> 1. The autoscaler monitor is frozen and "ray status" never updates.
> 2. The cluster never down-scales even when some nodes reach max idle time---It seems that the node statistic collection is also blocked.

1. is surprising because node creation happens in a background thread and thus shouldn't block the main autoscaler loop, unless the implementation is somehow causing a deadlock.
(The fact that the node creation happens in a background thread should be documented.)

2. There are a number of conditions that could lead to nodes not downscaling. This PR added a debug flag you can use to check if the failure of downscaling is due to Ray-internal logical resource usage: https://github.com/ray-project/ray/pull/22136

**Comment by DmitriGekhtman** [comment]
> a pending node is usually caused by the lack of resources in my cluster

One option is to just throw an error on create_node when there's a lack of resources in the cluster. This is what the underlying cloud API calls for GCPNodeProvider and AWSNodeProvider do when there aren't enough instances in the availability zone.

**Comment by TingkaiLiu** [comment]
> 1. is surprising because node creation happens in a background thread and thus shouldn't block the main autoscaler loop, unless the implementation is somehow causing a deadlock.
> (The fact that the node creation happens in a background thread should be documented.)

The deadlock assumption makes sense. When I was blocking create_node() I acquired the node status lock, which prevents other threads from reading the node status...my problem

**Comment by DmitriGekhtman** [comment]
> what is the expected behavior of IP query in node_provider for the nodes that don't have IP yet?

Should return "None".

**Comment by TingkaiLiu** [comment]
Several questions arise when further writing the doc:
1. For external_ip() function, the K8s node provider just raise an error saying "Must use internal IPs with K8s". In what case can this be allowed? It would be better if we can document how internal IP and external IP are used. 
2. In the function header of get_node_id(), it says "Assumes IP-address is unique per node." Is this necessary? It seems that autoscaler use node_ip to collect metrics from GCS, but I think GCS can differentiate different nodes with the same IP by port numbers. Is it just a design choice when writing the autoscaler?
3. For create_node() function, it says "Optionally returns a mapping from created node id to node metadata." What exactly is this "metadata" and how will it be used?
4. create_node_with_resource() says "prefer to implement this when possibly directly", which seems to conflict your comment above. Could be explain a little bit?
5. You mentioned bootstrap_config() can change the cluster config in arbitrary way, I think fillout_available_node_types_resources() can do that too?

**Comment by TingkaiLiu** [comment]
Another point regarding this:

> The "nodes" should be virtual machines that can be ssh-d into using the output of the internal_ip and external_ip methods.

This may not me necessary true, though. For K8s cluster it relies on the K8s command runner to run setup commands. For the customize platform which I am implementing node_provider for, SSH is also not the way to run commands. (This even leads to another topic: the command runner may not be fully needed if all the setup commands are bound to node creation, at least on the platform I am using.)

**Comment by DmitriGekhtman** [comment]
> Ok, that works too.

Yeah, the autoscaling config has to find some way to get onto the head node.

**Comment by DmitriGekhtman** [comment]
> 1. For external_ip() function, the K8s node provider just raise an error saying "Must use internal IPs with K8s". In what case can this be allowed? It would be better if we can document how internal IP and external IP are used.

K8s node provider is on the path to deprecation. external_ips are used in "Ray up" when starting the head node. internal_ips are used for communication within the cluster. I believe there's a "use_internal_ips" argument passed into the NodeUpdater that determines which ip is used to execute SSH commands.

> 2. In the function header of get_node_id(), it says "Assumes IP-address is unique per node." Is this necessary? It seems that autoscaler use node_ip to collect metrics from GCS, but I think GCS can differentiate different nodes with the same IP by port numbers. Is it just a design choice when writing the autoscaler?

Yes. One Ray node per VM per ip was the assumption when the autoscaler was implemented. As usual, there are ways of hacking around it, but it's recommended to stick to this paradigm if possible.

> 3. For create_node() function, it says "Optionally returns a mapping from created node id to node metadata." What exactly is this "metadata" and how will it be used?

The doc string should be improved. It's not used by the autoscaler. For certain external integrations or for testing, it might be useful to return data about the created nodes.

> 4. create_node_with_resource() says "prefer to implement this when possibly directly", which seems to conflict your comment above. Could be explain a little bit?

I might be in disagreement with the author of the doc string :)
`create_node_with_resources` is what you'd call if you're not using node updaters to start Ray and instead are using this call to create node, start Ray, and advertise the resources to Ray in the `ray start` command. This method was introduced for a testing local NodeProvider, I think called `FakeNodeProvider`. You can take a look at the implementation of that class to figure out how to use this. Given that you are running empty NodeUpdaters, this method might be useful for you.

> 5. You mentioned bootstrap_config() can change the cluster config in arbitrary way, I think fillout_available_node_types_resources() can do that too?

That is true. The method `fillout_available_node_types_resources` is meant to optionally auto-fill in the `resources` field of each node type based on config data (for example, the cloud instance type). You could take a look at the implementation for AWSNodeProvider to see how this works.

> 6. By the way, should I set up the document following https://docs.ray.io/en/latest/ray-contribute/docs.html?

Sure! I believe those instructions are up to date.

**Comment by TingkaiLiu** [comment]
I have submitted a PR:
https://github.com/ray-project/ray/pull/26715

Feel free to further edit the document or move it around in the doc structure!

**Comment by nadongjun** [comment]
> I have submitted a PR: [#26715](https://github.com/ray-project/ray/pull/26715)
> 
> Feel free to further edit the document or move it around in the doc structure!

This document has been very helpful to me. I think it would be great to add this content to the official documentation. Is there any plan to merge PR #26715?
